database/sqlite3.ksy

```diff
@@ -32,213 +32,431 @@ doc: |
+  btree_page_pointer:
+    seq:
+      - id: page_number
+        type: u4
+    instances:
+      page:
+        io: _root._io
+        pos: (page_number - 1) * _root.header.page_size
+        size: _root.header.page_size
+        type: btree_page(page_number)
+        if: page_number != 0
   btree_page:
+    params:
+      - id: page_number
+        type: u4
     seq:
+      - id: database_header
+        type: database_header
+        if: page_number == 1
       - id: page_type
         type: u1
+        enum: btree_page_type
       - id: first_freeblock
         type: u2
-      - id: num_cells
+        doc: The start of the first freeblock on the page, or is zero if there are no freeblocks.
+      - id: cell_count
         type: u2
-      - id: ofs_cells
+        doc: The number of cells on the page
+      - id: cell_content_area_start_raw
         type: u2
+        doc: |
+          The start of the cell content area. A zero value for this integer is interpreted as 65536.
+          The interpreted value is available as `cell_content_area`.
       - id: num_frag_free_bytes
         type: u1
+        doc: The number of fragmented free bytes within the cell content area.
       - id: right_ptr
-        type: u4
-        if: page_type == 2 or page_type == 5
+        type: btree_page_pointer
+        if: page_type == btree_page_type::index_interior or page_type == btree_page_type::table_interior
+        doc: |
+          The right-most pointer. This value appears in the header of interior
+          b-tree pages only and is omitted from all other pages.
       - id: cells
-        type: ref_cell
+        type: cell_pointer
         repeat: expr
-        repeat-expr: num_cells
-  ref_cell:
+        repeat-expr: cell_count
+    instances:
+      call_content_area_start:
+        value: 'cell_content_area_start_raw == 0 ? 65536 : cell_content_area_start_raw'
+      cell_content_area:
+        pos: call_content_area_start
+        size: _root.header.usable_size - call_content_area_start
+      reserved_area:
+        pos: _root.header.page_size - _root.header.reserved_space
+        size-eos: true
+        if: _root.header.reserved_space != 0
+  cell_pointer:
     seq:
-      - id: ofs_body
+      - id: content_offset
         type: u2
     instances:
-      body:
-        pos: ofs_body
+      content:
+        pos: content_offset
         type:
           switch-on: _parent.page_type
           cases:
-            0x0d: cell_table_leaf
-            0x05: cell_table_interior
-            0x0a: cell_index_leaf
-            0x02: cell_index_interior
-  cell_table_leaf:
+            btree_page_type::table_leaf: table_leaf_cell
+            btree_page_type::table_interior: table_interior_cell
+            btree_page_type::index_leaf: index_leaf_cell
+            btree_page_type::index_interior: index_interior_cell
+  table_leaf_cell:
     doc-ref: 'https://www.sqlite.org/fileformat.html#b_tree_pages'
     seq:
-      - id: len_payload
+      - id: p
```

Comment:
`p` is cryptic ... why not keep `len_payload`? or `payload_size`?

https://www.sqlite.org/fileformat.html

> Table B-Tree Leaf Cell (header 0x0d):
> 
> - A varint which is the total number of bytes of payload, including any overflow
> - A varint which is the integer key, a.k.a. "[rowid](https://www.sqlite.org/lang_createtable.html#rowid)"
> - The initial portion of the payload that does not spill to overflow pages.
> - A 4-byte big-endian integer page number for the first page of the overflow page list - omitted if all payload fits on the b-tree page.

Reply:
See my earlier comment. This was also also a reference to the formulas in the file format documentation. After restructuring the format a couple of times, retaining the short name doesn't make much sense anymore. I've changed it to `payload_size`.

Reply:
> Ideally use the same name-format for all sizes/lengths: payload.header_length, page.cell_count, ...

I've reworked the definition in an attempt to keep things consistent. `count` is a number of things, `size` is byte size, `length` is only used for blob and string now. I personally think it's clearer to keep count and size distinct.

Reply:
I ended up with `num_...` instead to silence the Kaitai compiler warnings. Compiles without warnings now.

Reply:
@pepijnve:

> I ended up with `num_...` instead to silence the Kaitai compiler warnings.

That's correct. It's preferred to use the naming convention as described in https://doc.kaitai.io/ksy_style_guide.html#attr-id. Before your changes, the spec tried to stick to them (`num_cells`, `ofs_cells`, `ofs_body`, `len_payload`, `len_page`), so I'm not sure why you felt the need to change them.

> Compiles without warnings now.

You shouldn't be only fixing the warnings (they are not complete, and don't work all the time) - please at least check out https://doc.kaitai.io/ksy_style_guide.html#attr-id and make sure you follow the recommendations. For example, offsets should be called `ofs_{structure}` - the compiler doesn't generate warnings for this yet.

Reply:
I tested the produced code with the Java backend on a database with a table b-tree that has both interior and leaf pages. Seems to work correctly. The root page is an interior page and has a non-null right pointer. If I follow that I end up at a leaf page with a null right pointer.

One pitfall to be aware of here is that page numbers are 1-based in the file format, while the pages list is 0-based?

In the Java implementation, the Kaitai compiler also generates an eager `ArrayList` for the `pages` instance. Kind of useless for large files.

Reply:
> In the Java implementation, the Kaitai compiler also generates an eager `ArrayList` for the `pages` instance. Kind of useless for large files.

Well, it works the way you implemented it. The `pages` instance itself is lazily parsed, but that doesn't mean the individual pages will automatically be lazy too. In this case, it's definitely possible to make them lazy, only there isn't the magic switch you'd like.

You can do for example something like this (each page `body` will be parsed lazily):

```ksy
instances:
  pages:
    type: page(_index + 1, header.page_size * _index)
    repeat: expr
    repeat-expr: header.page_count # note: this should be called `num_pages` per KS conventions

types:
  page:
    params:
      - id: page_number
        type: s4
      - id: ofs_body
        type: s4
    instances:
      body:
        pos: ofs_body
        size: _parent.header.page_size
        # type: ...
```

Reply:
> so I'm not sure why you felt the need to change them.

Potayto, potahto. What's legible to one person is gobbledygook to someone else. See for instance Microsoft's Hungarian notation. Personally I find a suffix stlye (`foobar_offset`) easier to read than a prefix style (`ofs_foobar`). The former I can read in the order it's written, the other asks for interpretation and flipping the word order.

I made this more complete definition for a SQLite internals sessions for my colleagues. It was mainly written with that in mind and since it was more complete and correct than what was already I simply made the PR with what I had at that point.
No objections to sticking to the conventions. Should be in line with what you asked for now.